### PR TITLE
Add methods on Branch to simplify data updates/next version ops

### PR DESCRIPTION
**Pull request by genfx999**
# Citrine Python PR

## Description 
Add methods directly to Branch object to make it easier to get updates and advance to next version of a branch

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Maintenance (non-breaking change to assist developers)

### Adherence to team decisions
- [x] I have added tests for 100% coverage
- [x] I have written Numpy-style docstrings for every method and class.
- [x] I have communicated the downstream consequences of the PR to others.
- [x] I have bumped the version in __version\__.py

### CODE REVIEW

File: src/citrine/resources/branch.py [hunk 1]
```diff
@@ -61,6 +61,59 @@ def experiment_datasource(self) -> Optional[ExperimentDataSource]:
...
+            scheduled for training.
+
+        Returns
+        -------
+        Branch
+            The new branch record after version update
+
+        """
+        path_template = (f'/projects/{{project_id}}/branches/'
+                         f'{{branch_id}}/data-version-updates-predictor')
```
Comment (anoto-moniz) [inline src/citrine/resources/branch.py:91]: The f-string and double braces cancel each other out. You can just do

```
path_template = ('/projects/{project_id}/branches/'
                 '{branch_id}/data-version-updates-predictor')
```

Comment (anoto-moniz) [reply src/citrine/resources/branch.py:91]: Or you could further simplify it by using the template above:

```
path_template = self._path_template +
                '/{branch_id}/data-version-updates-predictor'
```

File: src/citrine/resources/branch.py [hunk 2]
```diff
@@ -61,6 +61,59 @@ def experiment_datasource(self) -> Optional[ExperimentDataSource]:
...
+        path = format_escaped_url(path_template, project_id=self.project_id, branch_id=self.uid)
+        data = self.session.get_resource(path, version="v2")
+        version_updates = BranchDataUpdate.build(data)
+        # If no new data sources, then exit, nothing to do
+        if len(version_updates.data_updates) == 0:
+            return self
+
+        path_template = f'/projects/{{project_id}}/branches/next-version-predictor'
```
Comment (anoto-moniz) [inline src/citrine/resources/branch.py:99]: Similar to above: `path_template = f'{self._path_template}/next-version-predictor'`

File: src/citrine/resources/branch.py [hunk 3]
```diff
@@ -61,6 +61,59 @@ def experiment_datasource(self) -> Optional[ExperimentDataSource]:
         branch_erds_iter = erds.list(branch_id=self.uid, version="latest")
         return next(branch_erds_iter, None)
 
+    def update_data(self,
+                    *,
+                    use_existing=True,
+                    retrain_models=False):
```
Comment (anoto-moniz) [inline src/citrine/resources/branch.py:67]: Please add type hinting.

File: src/citrine/resources/branch.py [hunk 4]
```diff
@@ -61,6 +61,59 @@ def experiment_datasource(self) -> Optional[ExperimentDataSource]:
         branch_erds_iter = erds.list(branch_id=self.uid, version="latest")
         return next(branch_erds_iter, None)
 
+    def update_data(self,
```
Comment (anoto-moniz) [inline src/citrine/resources/branch.py:64]: The SDK pattern is for making requests to happen in the corresponding collection class.

**Review by anoto-moniz — CHANGES_REQUESTED** [review]
See the comments about moving `update_data` and adding type hints.

File: src/citrine/resources/branch.py (hunk 4 above)
Comment (genfx999) [reply src/citrine/resources/branch.py:64]: The request was for this call to be on the branch itself, I updated it to use the existing methods on the collection to do the actual requesting part.  Will that work?

Comment (anoto-moniz) [reply src/citrine/resources/branch.py:64]: I'll check with James. I'd like to preserve the existing pattern, but if he's dedicated to the method being on the `Branch` object itself, I'll give it the okay.

File: src/citrine/resources/branch.py [hunk 5]
```diff
@@ -61,6 +61,49 @@ def experiment_datasource(self) -> Optional[ExperimentDataSource]:
...
+                    *,
+                    use_existing: bool = True,
+                    retrain_models: bool = False):
```
Comment (anoto-moniz) [inline src/citrine/resources/branch.py:67]: Please add a type hint for the return type, too

**Review by anoto-moniz — APPROVED** [review]

**Pull request merged**